### PR TITLE
explain why finalizers are not processed in order

### PR DESCRIPTION
**Pull request by deads2k**
This came up in slack and appeared frequent enough to deserve a note for future us.

@kubernetes/sig-api-machinery-misc 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -250,6 +250,12 @@ type ObjectMeta struct {
 	// is an identifier for the responsible component that will remove the entry
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
+	// Finalizers may be processed and removed in any order.  Order is NOT enforced
+	// because it introduces significant risk of stuck finalizers.
+	// finalizers is a shared field, any actor with permission can reorder it.
+	// If the finalizer list is processed in order, then this can deadlock finalizers.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: "then this can deadlock finalizers" makes it sound like a bug in the API server. Maybe rephrase like "this can lead to a situation in which the component responsible for the first finalizer in the list is waiting for a condition produced by a component responsible for a finalizer later in the list, resulting in a deadlock."

Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: updated

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -250,6 +250,15 @@ type ObjectMeta struct {
 	// is an identifier for the responsible component that will remove the entry
 	// from the list. If the deletionTimestamp of the object is non-nil, entries
 	// in this list can only be removed.
+	// Finalizers may be processed and removed in any order.  Order is NOT enforced
+	// because it introduces significant risk of stuck finalizers.
+	// finalizers is a shared field, any actor with permission can reorder it.
```
Comment (seh) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]:
```suggestion
	// finalizers is a shared field; any actor with permission can reorder it.
```
Should the leading 'f' be capitalized? Is this documenting the Go field name, or the JSON field name?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -250,6 +250,15 @@ type ObjectMeta struct {
...
+	// If the finalizer list is processed in order, then this can lead to a situation
```
Comment (seh) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Carrying on from the previous sentence, consider starting this one to clarify:
> Since the finalizer order is difficult to coordinate, processing the finalizer list in order could lead to a situation in which the component responsible for the first finalizer in the list would be waiting for a signal (such as a particular field value to change or an external system to perform some operation) produced by a component responsible for a finalizer later in the list, resulting in a deadlock.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: documentation should reference the API field, so lowercase `f`

**Comment by liggitt** [comment]
verify shows generated file updates are required

**Comment by deads2k** [comment]
will be back to this soon. Reminder to self that I need to update this page: https://kubernetes.io/docs/reference/using-api/api-concepts/

**Comment by deads2k** [comment]
/hold

to stop verify retest
